src/conversions.rs

```diff
@@ -111,86 +228,195 @@ impl Optimizer {
-                }
-                Code::Label {
-                    pos: _pos,
-                    label: _label,
-                } => {
-                    // TODO handle labels
-                }
-            }
-        }
+        let codelist = block
+            .instrs
+            .iter()
+            .map(|instr| self.instr_to_code_expr(instr, &mut env))
+            .collect();
 
         Expr::Call(
-            "Func".into(),
+            "BlockNamed".into(),
             vec![
-                Expr::Lit(egglog::ast::Literal::String(func.name.clone().into())),
-                res,
+                Expr::Lit(egglog::ast::Literal::String(block.name.to_string().into())),
+                self.vec_to_codelist(codelist),
             ],
         )
     }
 
-    pub(crate) fn add_instr_effect(
+    pub(crate) fn instr_to_code_expr(
         &mut self,
         instr: &Instruction,
-        rest: &Expr,
-        env: &HashMap<String, Expr>,
+        env: &mut HashMap<String, Expr>,
     ) -> Expr {
-        match instr {
+        let (dest, expr) = match instr {
+            Instruction::Constant {
+                dest,
+                op: _op,
+                value,
+                pos: _pos,
+                const_type,
+            } => (
+                dest.clone(),
+                match value {
+                    Literal::Int(int) => Expr::Call(
+                        "Int".into(),
+                        vec![
+                            self.type_to_expr(const_type),
+                            Expr::Lit(egglog::ast::Literal::Int(*int)),
+                        ],
+                    ),
+                    Literal::Bool(bool) => {
+                        if *bool {
+                            Expr::Call("True".into(), vec![self.type_to_expr(const_type)])
+                        } else {
+                            Expr::Call("False".into(), vec![])
+                        }
+                    }
+                    Literal::Char(char) => Expr::Call(
+                        "Char".into(),
+                        vec![
+                            self.type_to_expr(const_type),
+                            Expr::Lit(egglog::ast::Literal::String(char.to_string().into())),
+                        ],
+                    ),
+                    Literal::Float(float) => Expr::Call(
+                        "Float".into(),
+                        vec![
+                            self.type_to_expr(const_type),
+                            Expr::Lit(egglog::ast::Literal::F64(OrderedFloat(*float))),
+                        ],
+                    ),
+                },
+            ),
+            Instruction::Value {
+                dest,
+                args,
+                funcs,
+                op,
+                labels: _labels,
+                pos: _pos,
+                op_type,
+            } => {
+                assert!(funcs.is_empty());
```

Comment:
why is funcs guaranteed empty here? can you add a comment?

Reply:
I think valid bril code should have funcs be empty for values, added a comment